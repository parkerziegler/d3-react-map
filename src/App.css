/** A constants SCSS partial for storing common variables and mixins */
@import url("https://fonts.googleapis.com/css?family=Roboto");
html,
body {
  margin: 0;
  height: 100%;
  width: 100%; }

.root {
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  align-items: center;
  min-height: 100%;
  min-width: 100%; }
  .root .page-container {
    display: flex;
    flex-direction: row;
    justify-content: space-between;
    align-items: center;
    text-align: center;
    min-width: 100%; }
    .root .page-container .page-content {
      display: flex;
      flex-direction: column;
      justify-content: center;
      align-items: center; }

.map-layout {
  display: flex;
  flex-direction: row; }

.chart-layout {
  display: flex;
  flex-direction: column;
  align-items: center;
  width: 100%;
  height: 100%; }

@media only screen and (max-width: 768px) {
  .map-layout {
    display: flex;
    flex-direction: column; } }
