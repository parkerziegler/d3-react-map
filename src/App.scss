@import './stylesheets/constants.scss';

body {
	@include flex-column;
	height: 100%;
	width: 100%;
	position: absolute;
	top: 0;
	left: 0;
	margin: 0;
}

#root {
	height: 100%;
	width: 100%;
}

.layout {
	@include flex-column;
	height: 100%;
	width: 100%;
	justify-content: space-around;
	align-items: center;

	.page-container {
		@include flex-row;
		height: 100%;
		width: 100%;
		justify-content: space-between;
		align-items: center;
		text-align: center;
	
		.page-content {
			display: flex;
			justify-content: center;
			align-items: center;
			max-height: 90%;
			max-width: 90%;
		}
	}
}

.percapita,
.raw {

	@include flex-row;
	width: 100%;
	height: auto;
	align-items: center;


	@media only screen and (max-width: 900px) {

		.map-container {
			width: 100%;
		}
	}
}

@media only screen and (max-width: 900px) {

	.percapita,
	.raw {
		@include flex-column;
	}
}
