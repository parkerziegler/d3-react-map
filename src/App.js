import React, { Component } from 'react';
import { connect } from 'react-redux';
import './App.css';
import Home from './components/home/Home';
import Map from './components/maps/Map/Map';
import MapDescription from './components/maps/MapDescription/MapDescription';
import BarChart from './components/graphs/BarChart/BarChart';
import Chevron from './components/navigation/Chevron/Chevron';
import { callAPI } from './actions/mapActions';
import PropTypes from 'prop-types';
import { Fragment } from 'redux-little-router';
import * as _ from 'lodash';

class App extends Component {

  constructor() {
    super();
    this.findKey = this.findKey.bind(this);
    this.getMainRoutes = this.getMainRoutes.bind(this);
    this.getChildRoutes = this.getChildRoutes.bind(this);
  }

  componentWillMount() {

    // on componentWillMount, invoke our callAPI saga to get our data
    const { dispatch } = this.props;
    dispatch(callAPI());
  }

  findKey(index) {

    // destructure props
    const { router } = this.props;

    // this is just a utility method to find the route index
    // that matches the one we passed in
    return _.findKey(router.routes, (route) => {
      return route.index === index;
    });
  }

  getMainRoutes() {

    // destructure props
    const { router } = this.props;

    let previousMainRoute;
    let nextMainRoute;

    // here we'll need to determine next and previous routes to pass
    // to our Chevron components
    let currentRoute = router.result;

    if (currentRoute.parent) {

      // this is a child route and has no main routes
      // just return '/'
      previousMainRoute = '/';
      nextMainRoute = '/';
    } else {

      // this is a main nav route
      // determine the previous and next routes

      // get the index of the previous main route
      let previousRouteIndex = currentRoute.index - 1;
    
      // add a check to see if we're on the initial route
      previousMainRoute = previousRouteIndex === -1 ? '/' : this.findKey(previousRouteIndex);

      // get the index of the next main route
      let nextRouteIndex = currentRoute.index + 1;

      // add a check to see if this is the last route
      nextMainRoute = currentRoute.isLastRoute ? '/' : this.findKey(nextRouteIndex);
    }

    return {
      previousMainRoute,
      nextMainRoute
    };

  }

  getChildRoutes() {

    const { router } = this.props;

    // here we need to determine previous and next routes
    // for the child navigation (i.e. within a map type)
    let currentRoute = router.result;

    let nextChildRoute;
    let previousChildRoute;

    // if the route has children, it is the parent
    // of the sub-navigation
    if (currentRoute.hasChildren) {

      // the next child route will be the first child
      // and its parent will have the same title as
      // the current route
      nextChildRoute = _.findKey(router.routes, (route) => {

        if (!route.parent) {
          return '/';
        }

        return route.parent.title === currentRoute.title && route.childIndex === 0;
      });

      // there is no previous child route for the parent
      previousChildRoute = '/';

    } else if (currentRoute.hasNextSibling) {

      // if the child route is bordered by other child routes,
      // implement similar logic to the main navigation
      let currentChildRouteIndex = currentRoute.childIndex;

      nextChildRoute = _.findKey(router.routes, (route) => {

        return route.parent.title === currentRoute.parent.title && route.childIndex === currentChildRouteIndex + 1;
      });

      previousChildRoute = currentChildRouteIndex === 0 ? currentRoute.parent.route : _.findKey(router.routes, (route) => {
        return route.parent.title === currentRoute.parent.title && route.childIndex === currentChildRouteIndex - 1;
      });

    } else if (currentRoute.isLastChildRoute) {

      // for the last child route, there is no next child
      let currentChildRouteIndex = currentRoute.childIndex;
      
      nextChildRoute = '/';

      // add a check to see if the previous route is the parent
      // if not, get the previous child route
      previousChildRoute = currentChildRouteIndex === 0 ? currentRoute.parent.route : _.findKey(router.routes, (route) => {
        return route.parent.title === currentRoute.parent.title && route.childIndex === currentChildRouteIndex - 1;
      });

    } else {

      // if this is not a child route, no need to worry about child navigation
      nextChildRoute = '/';
      previousChildRoute = '/';
    }

    return {
      previousChildRoute,
      nextChildRoute
    };

  }

  render() {

    // destructure props
    const { maps, router } = this.props;

    // obtain routes from getMainRoutes and getChildRoutes method
    let { previousMainRoute, nextMainRoute } = this.getMainRoutes();
    let { previousChildRoute, nextChildRoute } = this.getChildRoutes();

    // TODO - implement a Spinner solution
    let component = maps.fetchingData ?
      <div></div> :
      <Fragment forRoute={'/'}>
        <div className='layout'>
          <Chevron className='chevron-link top' path="M1683 1331l-166 165q-19 19-45 19t-45-19l-531-531-531 531q-19 19-45 19t-45-19l-166-165q-19-19-19-45.5t19-45.5l742-741q19-19 45-19t45 19l742 741q19 19 19 45.5t-19 45.5z" href={previousChildRoute} visible={previousChildRoute !== '/'} />
          <div className='page-container'>
            <Chevron className='chevron-link horizontal' path="M1427 301l-531 531 531 531q19 19 19 45t-19 45l-166 166q-19 19-45 19t-45-19l-742-742q-19-19-19-45t19-45l742-742q19-19 45-19t45 19l166 166q19 19 19 45t-19 45z" href={previousMainRoute} visible={previousMainRoute !== '/' || router.route === '/total-shootings'} />
            <div className='page-content'>
              <Fragment forRoute={'/'}>
                <Home />
              </Fragment>
              <Fragment forRoute={'/total-shootings'}>
                <div className='raw'>
                  <Map mapType='proportional' />
                  <MapDescription mapType='proportional' insetHeader='Total Shootings'/>
                </div>
              </Fragment>
              <Fragment forRoute={'/total-shootings/black'}></Fragment>
              <Fragment forRoute={'/total-shootings/latino'}></Fragment>
              <Fragment forRoute={'/total-shootings/asian'}></Fragment>
              <Fragment forRoute={'/total-shootings/nativeamerican'}></Fragment>
              <Fragment forRoute={'/total-shootings/white'}></Fragment>
              <Fragment forRoute={'/percapita'}>
                <div className='percapita'>
                  <Map mapType='choropleth' />
                  <MapDescription mapType='choropleth' insetHeader='Shootings per Million' />
                </div>
              </Fragment>
              <Fragment forRoute={'/shootingsbydate'}>
                  <BarChart />
              </Fragment>
            </div>
            <Chevron className='chevron-link horizontal' path="M1363 877l-742 742q-19 19-45 19t-45-19l-166-166q-19-19-19-45t19-45l531-531-531-531q-19-19-19-45t19-45l166-166q19-19 45-19t45 19l742 742q19 19 19 45t-19 45z" href={nextMainRoute} visible={nextMainRoute !== '/'} />
          </div>
          <Chevron className='chevron-link bottom' path="M1683 808l-742 741q-19 19-45 19t-45-19l-742-741q-19-19-19-45.5t19-45.5l166-165q19-19 45-19t45 19l531 531 531-531q19-19 45-19t45 19l166 165q19 19 19 45.5t-19 45.5z" href={nextChildRoute} visible={nextChildRoute !== '/'} nextSlide={router.routes[nextChildRoute].descSubtitle}/>
        </div>
      </Fragment>;

    return component;
  }
}

const mapStateToProps = (state) => {
  return {
    maps: state.mapReducer,
    router: state.router
  };
};

export default connect(mapStateToProps)(App);

// define propTypes for App
App.propTypes = {
  maps: PropTypes.object.isRequired,
  router: PropTypes.object.isRequired
};

