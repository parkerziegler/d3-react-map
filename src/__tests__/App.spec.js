import { App } from '../App';
import mockRouter from '../__mocks__/mock-router.json';

describe('<App />', () => {
  const props = {
    maps: {
      fetchingData: false,
    },
    router: mockRouter,
  };

  it('renders without crashing', () => {
    const wrapper = shallow(<App {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders an empty div when fetchingData is true', () => {
    const wrapper = shallow(<App {...props} maps={{ fetchingData: true }} />);
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('div')).toHaveLength(1);
  });

  it('renders correct chevron visibility across all routes', () => {
    const wrapper = shallow(
      <App
        {...props}
        router={{
          result: {
            index: 0,
          },
        }}
      />
    );
    expect(wrapper).toMatchSnapshot();

    wrapper.setProps({
      router: { result: { isLastRoute: true, parent: true } },
    });
    expect(wrapper).toMatchSnapshot();

    expect(wrapper.instance().getChevronVisibility('random')).toBe(false);
  });
});
