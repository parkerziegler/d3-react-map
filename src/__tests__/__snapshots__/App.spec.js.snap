// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> renders an empty div when fetchingData is true 1`] = `<div />`;

exports[`<App /> renders correct chevron visibility across all routes 1`] = `
<Connect(getContext(WithId))
  forRoute="/"
>
  <React.Fragment>
    <Chevron
      direction="up"
      onClick={[Function]}
      path="M1683 1331l-166 165q-19 19-45 19t-45-19l-531-531-531 531q-19 19-45 19t-45-19l-166-165q-19-19-19-45.5t19-45.5l742-741q19-19 45-19t45 19l742 741q19 19 19 45.5t-19 45.5z"
      visible={false}
    />
    <div
      className="page-container"
    >
      <Chevron
        direction="left"
        onClick={[Function]}
        path="M1427 301l-531 531 531 531q19 19 19 45t-19 45l-166 166q-19 19-45 19t-45-19l-742-742q-19-19-19-45t19-45l742-742q19-19 45-19t45 19l166 166q19 19 19 45t-19 45z"
        visible={false}
      />
      <div
        className="page-content"
      >
        <Connect(getContext(WithId))
          forRoute="/"
        >
          <Connect(Home) />
        </Connect(getContext(WithId))>
        <Connect(getContext(WithId))
          forRoute="/total-shootings"
        >
          <div
            className="map-layout"
          >
            <Connect(Map)
              mapType="proportional"
            />
            <Connect(MapDescription)
              mapType="proportional"
            />
          </div>
        </Connect(getContext(WithId))>
        <Connect(getContext(WithId))
          forRoute="/percapita"
        >
          <div
            className="map-layout"
          >
            <Connect(Map)
              mapType="choropleth"
            />
            <Connect(MapDescription)
              mapType="choropleth"
            />
          </div>
        </Connect(getContext(WithId))>
        <Connect(getContext(WithId))
          forRoute="/shootingsbydate"
        >
          <div
            className="chart-layout"
          >
            <Connect(Component) />
            <Connect(Line) />
          </div>
        </Connect(getContext(WithId))>
      </div>
      <Chevron
        direction="right"
        onClick={[Function]}
        path="M1363 877l-742 742q-19 19-45 19t-45-19l-166-166q-19-19-19-45t19-45l531-531-531-531q-19-19-19-45t19-45l166-166q19-19 45-19t45 19l742 742q19 19 19 45t-19 45z"
        visible={true}
      />
    </div>
    <Chevron
      direction="down"
      onClick={[Function]}
      path="M1683 808l-742 741q-19 19-45 19t-45-19l-742-741q-19-19-19-45.5t19-45.5l166-165q19-19 45-19t45 19l531 531 531-531q19-19 45-19t45 19l166 165q19 19 19 45.5t-19 45.5z"
      visible={false}
    />
  </React.Fragment>
</Connect(getContext(WithId))>
`;

exports[`<App /> renders correct chevron visibility across all routes 2`] = `
<Connect(getContext(WithId))
  forRoute="/"
>
  <React.Fragment>
    <Chevron
      direction="up"
      onClick={[Function]}
      path="M1683 1331l-166 165q-19 19-45 19t-45-19l-531-531-531 531q-19 19-45 19t-45-19l-166-165q-19-19-19-45.5t19-45.5l742-741q19-19 45-19t45 19l742 741q19 19 19 45.5t-19 45.5z"
      visible={true}
    />
    <div
      className="page-container"
    >
      <Chevron
        direction="left"
        onClick={[Function]}
        path="M1427 301l-531 531 531 531q19 19 19 45t-19 45l-166 166q-19 19-45 19t-45-19l-742-742q-19-19-19-45t19-45l742-742q19-19 45-19t45 19l166 166q19 19 19 45t-19 45z"
        visible={false}
      />
      <div
        className="page-content"
      >
        <Connect(getContext(WithId))
          forRoute="/"
        >
          <Connect(Home) />
        </Connect(getContext(WithId))>
        <Connect(getContext(WithId))
          forRoute="/total-shootings"
        >
          <div
            className="map-layout"
          >
            <Connect(Map)
              mapType="proportional"
            />
            <Connect(MapDescription)
              mapType="proportional"
            />
          </div>
        </Connect(getContext(WithId))>
        <Connect(getContext(WithId))
          forRoute="/percapita"
        >
          <div
            className="map-layout"
          >
            <Connect(Map)
              mapType="choropleth"
            />
            <Connect(MapDescription)
              mapType="choropleth"
            />
          </div>
        </Connect(getContext(WithId))>
        <Connect(getContext(WithId))
          forRoute="/shootingsbydate"
        >
          <div
            className="chart-layout"
          >
            <Connect(Component) />
            <Connect(Line) />
          </div>
        </Connect(getContext(WithId))>
      </div>
      <Chevron
        direction="right"
        onClick={[Function]}
        path="M1363 877l-742 742q-19 19-45 19t-45-19l-166-166q-19-19-19-45t19-45l531-531-531-531q-19-19-19-45t19-45l166-166q19-19 45-19t45 19l742 742q19 19 19 45t-19 45z"
        visible={false}
      />
    </div>
    <Chevron
      direction="down"
      onClick={[Function]}
      path="M1683 808l-742 741q-19 19-45 19t-45-19l-742-741q-19-19-19-45.5t19-45.5l166-165q19-19 45-19t45 19l531 531 531-531q19-19 45-19t45 19l166 165q19 19 19 45.5t-19 45.5z"
      visible={false}
    />
  </React.Fragment>
</Connect(getContext(WithId))>
`;

exports[`<App /> renders without crashing 1`] = `
<Connect(getContext(WithId))
  forRoute="/"
>
  <React.Fragment>
    <Chevron
      direction="up"
      onClick={[Function]}
      path="M1683 1331l-166 165q-19 19-45 19t-45-19l-531-531-531 531q-19 19-45 19t-45-19l-166-165q-19-19-19-45.5t19-45.5l742-741q19-19 45-19t45 19l742 741q19 19 19 45.5t-19 45.5z"
      visible={false}
    />
    <div
      className="page-container"
    >
      <Chevron
        direction="left"
        onClick={[Function]}
        path="M1427 301l-531 531 531 531q19 19 19 45t-19 45l-166 166q-19 19-45 19t-45-19l-742-742q-19-19-19-45t19-45l742-742q19-19 45-19t45 19l166 166q19 19 19 45t-19 45z"
        visible={true}
      />
      <div
        className="page-content"
      >
        <Connect(getContext(WithId))
          forRoute="/"
        >
          <Connect(Home) />
        </Connect(getContext(WithId))>
        <Connect(getContext(WithId))
          forRoute="/total-shootings"
        >
          <div
            className="map-layout"
          >
            <Connect(Map)
              mapType="proportional"
            />
            <Connect(MapDescription)
              mapType="proportional"
            />
          </div>
        </Connect(getContext(WithId))>
        <Connect(getContext(WithId))
          forRoute="/percapita"
        >
          <div
            className="map-layout"
          >
            <Connect(Map)
              mapType="choropleth"
            />
            <Connect(MapDescription)
              mapType="choropleth"
            />
          </div>
        </Connect(getContext(WithId))>
        <Connect(getContext(WithId))
          forRoute="/shootingsbydate"
        >
          <div
            className="chart-layout"
          >
            <Connect(Component) />
            <Connect(Line) />
          </div>
        </Connect(getContext(WithId))>
      </div>
      <Chevron
        direction="right"
        onClick={[Function]}
        path="M1363 877l-742 742q-19 19-45 19t-45-19l-166-166q-19-19-19-45t19-45l531-531-531-531q-19-19-19-45t19-45l166-166q19-19 45-19t45 19l742 742q19 19 19 45t-19 45z"
        visible={true}
      />
    </div>
    <Chevron
      direction="down"
      onClick={[Function]}
      path="M1683 808l-742 741q-19 19-45 19t-45-19l-742-741q-19-19-19-45.5t19-45.5l166-165q19-19 45-19t45 19l531 531 531-531q19-19 45-19t45 19l166 165q19 19 19 45.5t-19 45.5z"
      visible={true}
    />
  </React.Fragment>
</Connect(getContext(WithId))>
`;
